core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,394 @@
+        header: validator::BlockHeader {
+            parent: consensus.parent,
+            number,
+            payload: payload.hash(),
+        },
+        payload,
+        justification: consensus.justification,
+    })
+}
+
+async fn storage<'a>(
+    ctx: &ctx::Ctx,
+    pool: &'a ConnectionPool,
+) -> StorageResult<StorageProcessor<'a>> {
+    ctx.wait(pool.access_storage_tagged("sync_layer"))
+        .await?
+        .map_err(StorageError::Database)
+}
+
+async fn start_transaction<'a, 'b, 'c: 'b>(
+    ctx: &ctx::Ctx,
+    storage: &'c mut StorageProcessor<'a>,
+) -> Result<StorageProcessor<'b>, StorageError> {
+    ctx.wait(storage.start_transaction())
+        .await?
+        .map_err(|err| StorageError::Database(err.into()))
+}
+
+async fn commit<'a>(ctx: &ctx::Ctx, txn: StorageProcessor<'a>) -> StorageResult<()> {
+    ctx.wait(txn.commit())
+        .await?
+        .map_err(|err| StorageError::Database(err.into()))
+}
+
+async fn fetch_block<'a>(
+    ctx: &ctx::Ctx,
+    storage: &mut StorageProcessor<'a>,
+    number: validator::BlockNumber,
+) -> StorageResult<Option<validator::FinalBlock>> {
+    let number = MiniblockNumber(
+        number
+            .0
+            .try_into()
+            .context("MiniblockNumber")
+            .map_err(StorageError::Database)?,
+    );
+    let Some(block) = ctx
+        .wait(
+            storage
+                .sync_dal()
+                .sync_block(number, Address::default(), true),
+        )
+        .await?
+        .context("sync_block()")
+        .map_err(StorageError::Database)?
+    else {
+        return Ok(None);
+    };
+    if block.consensus.is_none() {
+        return Ok(None);
+    }
+    Ok(Some(
+        sync_block_to_consensus_block(block)
+            .context("sync_block_to_consensus_block()")
+            .map_err(StorageError::Database)?,
+    ))
+}
+
+async fn fetch_payload<'a>(
+    ctx: &ctx::Ctx,
+    storage: &mut StorageProcessor<'a>,
+    block_number: validator::BlockNumber,
+) -> StorageResult<Option<consensus::Payload>> {
+    let n = MiniblockNumber(
+        block_number
+            .0
+            .try_into()
+            .context("MiniblockNumber")
+            .map_err(StorageError::Database)?,
+    );
+    let Some(sync_block) = ctx
+        .wait(storage.sync_dal().sync_block(n, Address::default(), true))
+        .await?
+        .context("sync_block()")
+        .map_err(StorageError::Database)?
+    else {
+        return Ok(None);
+    };
+    Ok(Some(sync_block.try_into().map_err(StorageError::Database)?))
+}
+
+async fn put_block<'a>(
+    ctx: &ctx::Ctx,
+    storage: &mut StorageProcessor<'a>,
+    block: &validator::FinalBlock,
+) -> StorageResult<()> {
+    let n = MiniblockNumber(
+        block
+            .header
+            .number
+            .0
+            .try_into()
+            .context("MiniblockNumber")
+            .map_err(StorageError::Database)?,
+    );
+    let mut txn = start_transaction(ctx, storage)
```

Comment:
Just to understand better: Is there a risk here that a non-transactional update would result in a logical error, or is it just a best practice of wrapping related DB calls in a transaction?

Reply:
it is a read-write transaction and I don't want to rely on external assumptions about the storage access patterns if I don't have to.